File: frontend/types/public-voice-experiences.ts
export type PublicVoiceFieldType =
  | 'text'
  | 'textarea'
  | 'email'
  | 'phone'
  | 'integer'
  | 'select'
  | 'checkbox'
  | 'date';

export interface PublicVoiceFieldOption {
  value: string;
  label: string;
}

export interface PublicVoiceContextField {
  key: string;
  label: string;
  description: string | null;
  field_type: PublicVoiceFieldType;
  required: boolean;
  options: PublicVoiceFieldOption[];
}

export interface PublicVoiceExperience {
  slug: string;
  locale: string;
  version: number;
  content: {
    title: string;
    description: string;
    submit_label: string;
    call_label: string;
    success_message: string;
  };
  theme: {
    logo_url: string | null;
    primary_color: string | null;
    layout: 'centered' | 'split' | 'card';
  };
  consent: {
    required: boolean;
    label: string | null;
    privacy_url: string | null;
  };
  fields: PublicVoiceContextField[];
  call_settings: {
    auto_start: boolean;
    show_microphone_help: boolean;
    language: string;
  };
  capabilities: {
    submissions: true;
    calls: true;
  };
}

export type PublicSubmissionFieldErrorCode =
  | 'required'
  | 'unknown_field'
  | 'invalid_type'
  | 'too_long'
  | 'too_short'
  | 'invalid_option'
  | 'invalid_format'
  | 'consent_required';

export type PublicSubmissionErrorCode =
  | 'experience_version_changed'
  | 'validation_error'
  | 'verification_failed'
  | 'rate_limited'
  | 'internal_error';

export interface PublicSubmissionResponse {
  status: 'accepted';
  context_token: string;
  expires_at: string;
  capabilities: { submissions: true; calls: false };
}

export interface PublicSubmissionFailure {
  code: PublicSubmissionErrorCode;
  fields: Array<{ key: string; code: PublicSubmissionFieldErrorCode }>;
}

File: backend/app/schemas/public_voice_experiences.py
from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, ConfigDict, Field


class _PublicModel(BaseModel):
    model_config = ConfigDict(extra="forbid")


class PublicFieldOption(_PublicModel):
    value: str
    label: str


class PublicVoiceContextField(_PublicModel):
    key: str
    label: str
    description: str | None
    field_type: Literal[
        "text", "textarea", "email", "phone", "integer", "select", "checkbox", "date"
    ]
    required: bool
    options: list[PublicFieldOption]


class PublicVoiceContent(_PublicModel):
    title: str = Field(min_length=1, max_length=160)
    description: str = Field(min_length=1, max_length=2000)
    submit_label: str = Field(min_length=1, max_length=80)
    call_label: str = Field(min_length=1, max_length=80)
    success_message: str = Field(min_length=1, max_length=1000)


class PublicVoiceTheme(_PublicModel):
    logo_url: str | None
    primary_color: str | None = Field(None, pattern=r"^#[0-9A-Fa-f]{6}$")
    layout: Literal["centered", "split", "card"]


class PublicVoiceConsent(_PublicModel):
    required: bool
    label: str | None
    privacy_url: str | None


class PublicCapabilities(_PublicModel):
    submissions: bool = True
    calls: bool = True


class PublicVoiceCallSettings(_PublicModel):
    auto_start: bool = False
    show_microphone_help: bool = True
    language: str = Field(default="es", min_length=2, max_length=16)


class PublicVoiceExperienceResponse(_PublicModel):
    slug: str
    locale: str
    version: int
    content: PublicVoiceContent
    theme: PublicVoiceTheme
    consent: PublicVoiceConsent
    fields: list[PublicVoiceContextField]
    call_settings: PublicVoiceCallSettings
    capabilities: PublicCapabilities
